db/column_family.cc

```diff
@@ -699,19 +699,6 @@ ColumnFamilyData::~ColumnFamilyData() {
           id_, name_.c_str());
     }
   }
-
-  if (data_dirs_.size()) {  // Explicitly close data directories
```

Comment:
Thanks a lot for catching this bug! @pdillinger I wonder if maintaining an unordered_map between data directories and the ref counter could solve this? Here we just decrease the ref counter for each data directory and close the data directory when the ref counter is 0? Is this feasible?

Reply:
I should have mentioned: this PR is intended for patching previous releases. For that we want a kind of minimum sufficient fix.

And I authored this PR myself because I got most of the way here to help me understand the problem. Proper treatment of the refcounting stuff should (IMHO) incorporate both of my TODOs, so it's not a simple change and I don't intend to follow-up on it myself any time soon.

I'm not even really convinced it's that useful to ensure we always call Close. The practical effect of the complete ambiguity in the original API definition is that implementors need to effect a Close on destruction if it hasn't been called already. What do we gain by enforcing a stronger constraint on ourselves? And I'm not sure what kind of useful failure FSDirectory::Close would return to us in any environment. There can't (at present) be any pending buffered operations that might have failed in the background. If the intent was to future-proof the API in case such operations were added in the future, then Close should have been marked and enforced as required. At present, we don't even enforce Close on DB, which can return background errors like "your entire DB disappeared! It's probably not going to be here when you try to re-open!"

Reply:
I may follow up this later to refine the implementation and your point makes sense to me. I will take them into account and discuss with you later if I have something else proposed. As we are aiming for sufficient minimum changes, this PR looks good to me.

Reply:
> I'm not even really convinced it's that useful to ensure we always call Close. The practical effect of the complete ambiguity in the original API definition is that implementors need to effect a Close on destruction if it hasn't been called already. What do we gain by enforcing a stronger constraint on ourselves?

It's a good question. The original goal of #10049 was to release file descriptors when `DB::Close()` is called. It looks like it could've been achieved by clearing `DBImpl::data_dirs_` and resetting `DBImpl::{db,wal}_dir_` during `DB::Close()`.